libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/internal/MapboxNavigationOwner.kt

```diff
@@ -0,0 +1,59 @@
+package com.mapbox.navigation.ui.lifecycle.internal
+
+import android.app.Application
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.ui.lifecycle.MapboxNavigationInitializer
+import com.mapbox.navigation.ui.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.utils.internal.logI
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class MapboxNavigationOwner private constructor(
+    private val application: Application,
+    private val mapboxNavigationInitializer: MapboxNavigationInitializer
+) {
+    private val services = CopyOnWriteArraySet<MapboxNavigationObserver>()
+    private var mapboxNavigation: MapboxNavigation? = null
+    internal val carAppLifecycleObserver = object : DefaultLifecycleObserver {
+
+        override fun onStart(owner: LifecycleOwner) {
+            logI(TAG, Message("navigation.onStart"))
+            val navigationOptions = mapboxNavigationInitializer.create(application)
+            check(!MapboxNavigationProvider.isCreated()) {
+                "MapboxNavigation should only be destroyed by the MapboxNavigationLifetime"
+            }
+            val mapboxNavigation = MapboxNavigationProvider.create(navigationOptions)
+            this@MapboxNavigationOwner.mapboxNavigation = mapboxNavigation
+            services.forEach { it.onAttached(mapboxNavigation) }
+        }
+
+        override fun onStop(owner: LifecycleOwner) {
+            logI(TAG, Message("navigation.onStop"))
+            services.forEach { it.onDetached(mapboxNavigation!!) }
+            MapboxNavigationProvider.destroy()
+        }
+    }
+
+    fun register(mapboxNavigationObserver: MapboxNavigationObserver) = apply {
+        mapboxNavigation?.let { mapboxNavigationObserver.onAttached(it) }
+        services.add(mapboxNavigationObserver)
+    }
+
+    fun unregister(mapboxNavigationObserver: MapboxNavigationObserver) {
+        mapboxNavigationObserver.onDetached(mapboxNavigation)
```

Comment:
I don’t think we need to call onDetached, if mapboxNavigation is null. I would expect every onDetached call to follow exactly one onAttached call. And the argument should be non-null.

Reply:
Yeah I've gone back and forth on this but have also decided to remove the nullable. Over communicating the reasoning here: Having a nullable, supports construction initialization and removal. Removing the nullable, makes the construction initialization impossible. So the first approach opted for the extra signal. But am in the process of removing, because of the examples below

Supporting this pattern
``` kotlin
object : MapboxNavigationObserver {
  // FooReader is constructed with an enabled state
  val fooReader = FooReader()

  fun onAttached(mapboxNavigation: MapboxNavigation) {
    // -- snip -- //
  }

  fun onDetached(mapboxNavigation: MapboxNavigation?) {
    fooReader.disable()
  }
}
```

Removing, requires this pattern. Which I'm agreeing is better
``` kotlin
object : MapboxNavigationObserver {
  // FooReader is constructed with an enabled state
  late init var fooReader: FooReader

  fun onAttached(mapboxNavigation: MapboxNavigation) {
     fooReader = FooReader()
  }

  fun onDetached(mapboxNavigation: MapboxNavigation) {
    // This pattern will crash when MapboxNavigation is nullable
    fooReader.disable()
  }
}
```